cpp/src/io/csv/csv_writer.cu

```diff
@@ -0,0 +1,284 @@
+        NVStrings::destroy(rtn);
+        rtn = dstr;
+    }
+    return rtn;
+}
+
+//
+// The args structure is interpretted as documented in io_types.h
+// This will create a CSV format by allocating host memory for the
+// entire output and determine pointers for each row/column entry.
+// Each column is converted to an NVStrings instance and then
+// copied into their position in the output memory. This way,
+// one column is processed at a time minimizing device memory usage.
+//
+gdf_error write_csv(csv_write_arg* args)
+{
+    gdf_error rc = gdf_error::GDF_SUCCESS;
+
+    gdf_column** columns = args->data;
+    unsigned int count = (unsigned int)args->num_cols;
+    unsigned int rows = (unsigned int)columns[0]->size;
+    const char* filepath = args->filepath;
+    char delimiter[2] = {',','\0'};
+    if( args->delimiter )
+        delimiter[0] = args->delimiter;
+    const char* terminator = "\n";
+    if( args->line_terminator )
+        terminator = args->line_terminator;
+    const char* true_value = (args->true_value ? args->true_value : "true");
+    const char* false_value = (args->false_value ? args->false_value : "false");
+
+    // check for issues here
+    if( filepath==0 )
+    {
+        std::cerr << "write_csv: filepath not specified\n";
+        return GDF_INVALID_API_CALL;
+    }
+    if( count==0 )
+    {
+        std::cerr << "write_csv: num_cols is required\n";
+        return GDF_INVALID_API_CALL;
+    }
+    if( columns==0 )
+    {
+        std::cerr << "write_csv: invalid data values\n";
+        return GDF_INVALID_API_CALL;
+    }
+
+    // check all columns are the same size
+    for( int idx=0; idx < args->num_cols; ++idx )
+    {
+        gdf_column* col = args->data[idx];
+        NVStrings* strs = nullptr;
+        //printf("%d: %p %s %d %d\n", idx, col, col->col_name, (int)col->dtype, (int)col->size);
+        switch( col->dtype )
+        {
+            case GDF_INT8:
+            case GDF_INT32:
+            case GDF_INT64:
+            case GDF_FLOAT32:
+            case GDF_FLOAT64:
+            case GDF_DATE64:
+                if( (unsigned int)col->size != rows )
+                {
+                    std::cerr << "write_csv: columns sizes do not match (" << (int)col->size << "!=" << rows << ")\n";
+                    return GDF_COLUMN_SIZE_MISMATCH;
+                }
+                break;
+            case GDF_STRING:
+                strs = (NVStrings*)col->data;
+                if( strs->size() != rows )
+                {
+                    std::cerr << "write_csv: columns sizes do not match (" << strs->size() << "!=" << rows << ")\n";
+                    return GDF_COLUMN_SIZE_MISMATCH;
+                }
+                break;
+            default:
+                std::cerr << "write_csv: unknown column type: " << (int)col->dtype << "\n";
+                return GDF_UNSUPPORTED_DTYPE;
+        }
+    }
+
+    // check the file can be written
+    FILE* fh = fopen(filepath,"wb");
```

Comment:
Since this is new code, I'd like to avoid inheriting more C code. As such, I'd really prefer this to be done using C++ file IO, i.e., use `ofstream`